File: pruefstuecke/telefondienst.mjs
/**
 * Prüfstück: Der Telefondienst sagt in den Einstellungen, wie es steht.
 *
 * DER ANLASS ist eine Ansage des Inhabers vom 30.08.2026, mit Blick auf
 * die Webex-Bildschirmfotos: „Telefondienst Status soll auch in
 * Einstellungen sein."
 *
 * VORGEFUNDEN: Der Anmeldezustand stand NUR oben in der Markenleiste,
 * und er stand dort als das, was Liblinphone durchreicht. Bei einem
 * Fehlschlag las der Kunde „Anmeldung fehlgeschlagen: Forbidden" – ein
 * englisches Wort aus einem SIP-Protokoll. Wer das liest, weiß nicht,
 * ob sein Passwort falsch ist, die Nebenstelle nicht existiert oder das
 * Netz klemmt. Genau in diesem Moment ruft er an.
 *
 * WAS HIER FESTGENAGELT WIRD:
 *
 *  1. Der Zustand steht in den EINSTELLUNGEN, nicht nur in der Leiste.
 *     Dort sucht man ihn, wenn etwas nicht geht.
 *  2. Der Grund steht auf DEUTSCH und nennt die Ursache, nicht den
 *     SIP-Code. „Benutzername oder Passwort stimmt nicht" ist eine
 *     Auskunft; „Forbidden" ist keine.
 *  3. Der Zustand hängt nicht an der FARBE allein. Wer Rot und Grün
 *     nicht unterscheidet, liest den Text – dieselbe Regel wie in der
 *     Markenleiste.
 *  4. Es gibt einen Weg, sich neu anzumelden, ohne das Passwort noch
 *     einmal einzutippen.
 *
 * Aufruf:  node pruefstuecke/telefondienst.mjs
 */
import { readFileSync, existsSync, readdirSync } from "node:fs";

const ergebnisse = [];
const pruefe = (name, ok, extra = "") => {
  ergebnisse.push([name, ok]);
  console.log((ok ? "  ok   " : " FEHL  ") + name + (ok || !extra ? "" : "  – " + extra));
};

const QUELLEN = "app/src/main/java/de/dariatech/softphone";
const LAYOUT = "app/src/main/res/layout/activity_main.xml";
const TEXTE = "app/src/main/res/values/strings.xml";
const dateien = existsSync(QUELLEN) ? readdirSync(QUELLEN).filter((d) => d.endsWith(".kt")) : [];
const lies = (d) => readFileSync(`${QUELLEN}/${d}`, "utf-8");
const layout = existsSync(LAYOUT) ? readFileSync(LAYOUT, "utf-8") : "";
const texte = existsSync(TEXTE) ? readFileSync(TEXTE, "utf-8") : "";

console.log("\n1) Es gibt eine Stelle, die den Zustand übersetzt");
{
  pruefe("Telefondienst.kt liegt vor", dateien.includes("Telefondienst.kt"),
    "hier gehört die Übersetzung von SIP nach Deutsch hin");
}
const dienst = dateien.includes("Telefondienst.kt") ? lies("Telefondienst.kt") : "";

console.log("\n2) Jeder Zustand hat einen deutschen Namen");
{
  for (const wort of ["Verbunden", "Nicht verbunden", "Wird angemeldet"]) {
    pruefe(`der Zustand „${wort}" ist benannt`, dienst.includes(wort));
  }
  pruefe("und der Fall ohne Zugangsdaten ist eigens benannt",
    /noch nicht eingerichtet/i.test(dienst),
    "wer nichts eingetragen hat, hat keine Störung – er hat noch nichts eingetragen");
}

console.log("\n3) Der Grund steht auf Deutsch, nicht als SIP-Wort");
{
  // Die Fehlschläge, die im Betrieb wirklich vorkommen. Jeder braucht
  // eine Auskunft, aus der hervorgeht, WAS zu tun ist.
  const faelle = [
    ["Unauthorized", "falsches Passwort"],
    ["Forbidden", "Anlage weist ab"],
    ["Not Found", "Nebenstelle gibt es nicht"],
    ["Timeout", "Anlage antwortet nicht"],
    ["Service Unavailable", "Anlage nimmt nichts an"]
  ];
  // OHNE RÜCKSICHT AUF GROSS UND KLEIN. Der erste Anlauf verglich
  // zeichengenau und meldete fünf Fehlschläge, obwohl der Code alle
  // fünf Fälle abdeckte – er vergleicht die kleingeschriebene Meldung,
  // weil Liblinphone mal „Forbidden" und mal „403 forbidden" liefert.
  // Ein Prüfstück, das an der Schreibweise scheitert, prüft die
  // Schreibweise und nicht die Sache.
  for (const [sip, wofuer] of faelle) {
    pruefe(`${sip} wird übersetzt (${wofuer})`,
      dienst.toLowerCase().includes(sip.toLowerCase()),
      "sonst liest der Kunde das englische SIP-Wort");
  }
  pruefe("und es gibt einen Rückfall für alles Übrige",
    /else ->|return .*meldung|sonst/i.test(dienst),
    "eine unbekannte Meldung darf nicht in einer leeren Zeile enden");
}

console.log("\n4) Die Auskunft kommt nie an ein Geheimnis heran");
{
  // Eine Statusmeldung wird vorgezeigt, abfotografiert und in eine
  // E-Mail an den Kundendienst geklebt. Sie darf kein Geheimnis tragen.
  //
  // DER ERSTE ANLAUF PRÜFTE DAS FALSCHE: Er verbot das WORT „Passwort"
  // in der Datei – und wurde rot an dem Satz „Benutzername oder
  // Passwort stimmt nicht", also genau an der Auskunft, um die es
  // geht. Verboten gehört nicht das Wort, sondern der ZUGRIFF: Diese
  // Datei darf nirgends an den gespeicherten Wert herankommen.
  const zugriff = [];
  for (const m of dienst.matchAll(/Zugangsspeicher\.\w+|getString\(\s*"password"|authInfo/g)) {
    zugriff.push(m[0]);
  }
  pruefe("Telefondienst.kt liest nirgends ein Geheimnis",
    zugriff.length === 0, zugriff.join(", "));
}

console.log("\n5) In den Einstellungen steht er wirklich");
{
  for (const [feld, name] of [
    ["@+id/dienstPunkt", "ein Punkt für den Zustand"],
    ["@+id/dienstText", "und der Zustand als TEXT daneben"],
    ["@+id/dienstGrund", "eine Zeile für den Grund"],
    ["@+id/dienstNebenstelle", "die eigene Nebenstelle"],
    ["@+id/neuAnmelden", "ein Knopf zum erneuten Anmelden"]
  ]) {
    pruefe(name, layout.includes(feld), feld);
  }
  pruefe("der Abschnitt ist überschrieben", /telefondienst_section/.test(texte));
}

console.log("\n6) Der Zustand kommt auch dort an");
{
  const haupt = dateien.includes("MainActivity.kt") ? lies("MainActivity.kt") : "";
  pruefe("onRegistration versorgt die Einstellungen",
    /dienstText|zeigeDienst/.test(haupt),
    "sonst steht dort für immer der Zustand vom Start");
  pruefe("der Knopf meldet wirklich neu an",
    /neuAnmelden\.setOnClickListener/.test(haupt));
  pruefe("und der Stack kann das auch",
    dateien.includes("LinphoneManager.kt") &&
      /refreshRegisters|neuAnmelden/.test(lies("LinphoneManager.kt")),
    "ohne refreshRegisters passiert beim Drücken nichts");
}

const fehler = ergebnisse.filter(([, ok]) => !ok);
console.log(`\n${ergebnisse.length - fehler.length}/${ergebnisse.length} Prüfungen bestanden`);
if (fehler.length > 0) console.log("FEHLGESCHLAGEN: " + fehler.map(([n]) => n).join(", "));
process.exit(fehler.length > 0 ? 1 : 0);

File: app/src/main/java/de/dariatech/softphone/Telefondienst.kt
package de.dariatech.softphone

import org.linphone.core.RegistrationState

/**
 * Der Zustand des Telefondienstes – in Worten, die jemand versteht.
 *
 * DER ANLASS ist eine Ansage des Inhabers vom 30.08.2026: Der Zustand
 * des Telefondienstes gehört in die Einstellungen, nicht nur in die
 * Leiste oben.
 *
 * WAS VORHER STAND: Bei einem Fehlschlag reichte die App durch, was
 * Liblinphone lieferte – „Anmeldung fehlgeschlagen: Forbidden". Das ist
 * ein englisches Wort aus einem SIP-Protokoll. Wer es liest, weiß
 * nicht, ob sein Passwort falsch ist, die Nebenstelle nicht existiert
 * oder das Netz klemmt. Und er liest es genau in dem Moment, in dem er
 * telefonieren will.
 *
 * DIE ÜBERSETZUNG IST DER EIGENTLICHE INHALT dieser Datei. Ein Zustand
 * ohne Grund ist eine halbe Auskunft; ein Grund im Fachjargon ist gar
 * keine.
 *
 * Dasselbe gibt es auf iOS (`Telefondienst.swift`) mit denselben
 * Sätzen. Zwei Apps, die verschieden erklären, erzeugen zwei
 * Fehlerbilder für dieselbe Ursache.
 */
enum class Dienstzustand {
    /** Angemeldet, die Anlage nimmt Anrufe an. */
    VERBUNDEN,

    /** Die Anmeldung läuft gerade. */
    LAEUFT,

    /** Angemeldet werden konnte nicht – der Grund steht daneben. */
    GETRENNT,

    /**
     * Es wurden noch keine Zugangsdaten eingetragen.
     *
     * EIN EIGENER ZUSTAND, kein „getrennt": Wer nichts eingetragen hat,
     * hat keine Störung. Ihm ein rotes „Nicht verbunden" hinzustellen,
     * schickt ihn auf Fehlersuche statt in das Feld darunter.
     */
    UNVOLLSTAENDIG
}

object Telefondienst {

    /** Der Zustand als Wort – nie nur als Farbe. */
    fun bezeichnung(zustand: Dienstzustand): String = when (zustand) {
        Dienstzustand.VERBUNDEN -> "Verbunden"
        Dienstzustand.LAEUFT -> "Wird angemeldet …"
        Dienstzustand.GETRENNT -> "Nicht verbunden"
        Dienstzustand.UNVOLLSTAENDIG -> "Noch nicht eingerichtet"
    }

    fun zustandAus(state: RegistrationState?): Dienstzustand = when (state) {
        RegistrationState.Ok -> Dienstzustand.VERBUNDEN
        RegistrationState.Progress -> Dienstzustand.LAEUFT
        else -> Dienstzustand.GETRENNT
    }

    /**
     * Aus der SIP-Meldung einen Satz machen, der weiterhilft.
     *
     * Geprüft wird auf Bestandteile und nicht auf Gleichheit: Was
     * Liblinphone durchreicht, ist mal „Forbidden", mal
     * „403 Forbidden", mal ein Satz mit dem Wort darin. Ein Vergleich
     * auf Gleichheit hätte in genau dem Fall danebengegriffen, in dem
     * die Auskunft gebraucht wird.
     *
     * DIE ORIGINALMELDUNG BLEIBT ERHALTEN, wo sie unbekannt ist – eine
     * Meldung, die niemand übersetzt hat, ist immer noch mehr wert als
     * eine leere Zeile. Der Kundendienst kann damit arbeiten.
     */
    fun grund(meldung: String): String {
        val m = meldung.lowercase()
        return when {
            m.contains("unauthorized") || m.contains("401") ->
                "Benutzername oder Passwort stimmt nicht."
            m.contains("forbidden") || m.contains("403") ->
                "Die Anlage weist diese Anmeldung ab. Meist ist die Nebenstelle " +
                    "gesperrt oder für ein anderes Gerät vergeben."
            m.contains("not found") || m.contains("404") ->
                "Diese Nebenstelle gibt es auf der Anlage nicht."
            m.contains("timeout") || m.contains("408") ->
                "Die Anlage antwortet nicht. Meist liegt es am Netz – im " +
                    "WLAN kann auch ein Router mit SIP-Hilfe (SIP-ALG) dazwischenfunken."
            m.contains("service unavailable") || m.contains("503") ->
                "Die Anlage nimmt gerade keine Anmeldungen an. Später erneut versuchen."
            m.contains("io error") || m.contains("no route") || m.contains("unreachable") ->
                "Keine Verbindung zur Anlage. Mobilfunk oder WLAN prüfen."
            meldung.isBlank() -> ""
            else -> meldung
        }
    }
}
